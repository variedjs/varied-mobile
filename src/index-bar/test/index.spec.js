import { mount, trigger, triggerDrag, mockScrollIntoView } from "../../../test";
import indexAnchor from "../../index-anchor";
import indexBar from "../";

test("should allow to custom anchor text", () => {
  const wrapper = mount(
    {
      template: `
      <vm-index-bar>
        <vm-index-anchor index="A">Title A</vm-index-anchor>
        <vm-index-anchor index="B">Title B</vm-index-anchor>
      </vm-index-bar>
    `
    },
    {
      stubs: {
        "vm-index-bar": indexBar,
        "vm-index-anchor": indexAnchor
      }
    }
  );

  expect(wrapper).toMatchSnapshot();
});

test("should scroll to anchor and emit select event after clicking the index-bar", () => {
  const onSelect = jest.fn();
  const wrapper = mount(
    {
      template: `
      <vm-index-bar @select="onSelect">
        <vm-index-anchor index="A" />
        <vm-index-anchor index="B" />
      </vm-index-bar>
    `,
      methods: {
        onSelect
      }
    },
    {
      stubs: {
        "vm-index-bar": indexBar,
        "vm-index-anchor": indexAnchor
      }
    }
  );

  const fn = mockScrollIntoView();
  const indexes = wrapper.findAll(".vm-index-bar-index");
  indexes.at(0).trigger("click");

  expect(fn).toHaveBeenCalledTimes(1);
  expect(onSelect).toHaveBeenCalledWith("A");
});

test("should scroll to anchor after touching the index-bar", () => {
  const onSelect = jest.fn();
  const wrapper = mount(
    {
      template: `
      <vm-index-bar @select="onSelect">
        <vm-index-anchor index="A" />
        <vm-index-anchor index="B" />
        <vm-index-anchor index="XXX" />
      </vm-index-bar>
    `,
      methods: {
        onSelect
      }
    },
    {
      stubs: {
        "vm-index-bar": indexBar,
        "vm-index-anchor": indexAnchor
      }
    }
  );

  const fn = mockScrollIntoView();
  const sidebar = wrapper.find(".vm-index-bar-sidebar");
  const indexes = wrapper.findAll(".vm-index-bar-index");

  document.elementFromPoint = function(x, y) {
    const index = y / 100;

    if (index === 1 || index === 2) {
      return indexes.at(index).element;
    }

    if (index === 3) {
      return {
        dataset: {}
      };
    }
  };

  // horizontal drag
  triggerDrag(sidebar, 100, 0);
  expect(fn).toHaveBeenCalledTimes(0);

  // vertical drag
  trigger(sidebar, "touchstart", 0, 0);
  trigger(sidebar, "touchmove", 0, 100);
  trigger(sidebar, "touchmove", 0, 200);
  trigger(sidebar, "touchmove", 0, 300);
  trigger(sidebar, "touchmove", 0, 400);
  trigger(sidebar, "touchend", 0, 400);
  expect(fn).toHaveBeenCalledTimes(1);
  expect(onSelect).toHaveBeenCalledWith("B");
});

test("should update active anchor after page scroll", () => {
  const nativeRect = Element.prototype.getBoundingClientRect;
  Element.prototype.getBoundingClientRect = function() {
    const { index } = this.dataset;
    return {
      top: index ? index * 10 : 0,
      height: 10
    };
  };

  const wrapper = mount(
    {
      template: `
      <vm-index-bar :sticky="sticky">
        <vm-index-anchor
          v-for="index in 4"
          :key="index"
          :index="index"
          :data-index="index - 1"
        />
      </vm-index-bar>
    `,
      data() {
        return {
          sticky: false
        };
      }
    },
    {
      stubs: {
        "vm-index-bar": indexBar,
        "vm-index-anchor": indexAnchor
      }
    }
  );

  window.scrollTop = 0;
  trigger(window, "scroll");
  expect(wrapper).toMatchSnapshot();

  wrapper.setData({ sticky: true });
  trigger(window, "scroll");
  expect(wrapper).toMatchSnapshot();
  wrapper.vm.$destroy();

  Element.prototype.getBoundingClientRect = nativeRect;
});
