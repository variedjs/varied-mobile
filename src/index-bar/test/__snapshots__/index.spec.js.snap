// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow to custom anchor text 1`] = `
<div class="vm-index-bar">
  <div class="vm-index-bar-sidebar"><span data-index="A" class="vm-index-bar-index">
      A
    </span><span data-index="B" class="vm-index-bar-index">
      B
    </span><span data-index="C" class="vm-index-bar-index">
      C
    </span><span data-index="D" class="vm-index-bar-index">
      D
    </span><span data-index="E" class="vm-index-bar-index">
      E
    </span><span data-index="F" class="vm-index-bar-index">
      F
    </span><span data-index="G" class="vm-index-bar-index">
      G
    </span><span data-index="H" class="vm-index-bar-index">
      H
    </span><span data-index="I" class="vm-index-bar-index">
      I
    </span><span data-index="J" class="vm-index-bar-index">
      J
    </span><span data-index="K" class="vm-index-bar-index">
      K
    </span><span data-index="L" class="vm-index-bar-index">
      L
    </span><span data-index="M" class="vm-index-bar-index">
      M
    </span><span data-index="N" class="vm-index-bar-index">
      N
    </span><span data-index="O" class="vm-index-bar-index">
      O
    </span><span data-index="P" class="vm-index-bar-index">
      P
    </span><span data-index="Q" class="vm-index-bar-index">
      Q
    </span><span data-index="R" class="vm-index-bar-index">
      R
    </span><span data-index="S" class="vm-index-bar-index">
      S
    </span><span data-index="T" class="vm-index-bar-index">
      T
    </span><span data-index="U" class="vm-index-bar-index">
      U
    </span><span data-index="V" class="vm-index-bar-index">
      V
    </span><span data-index="W" class="vm-index-bar-index">
      W
    </span><span data-index="X" class="vm-index-bar-index">
      X
    </span><span data-index="Y" class="vm-index-bar-index">
      Y
    </span><span data-index="Z" class="vm-index-bar-index">
      Z
    </span></div>
  <div>
    <div class="vm-index-anchor">Title A</div>
  </div>
  <div>
    <div class="vm-index-anchor">Title B</div>
  </div>
</div>
`;

exports[`should update active anchor after page scroll 1`] = `
<div class="vm-index-bar">
  <div class="vm-index-bar-sidebar"><span data-index="A" class="vm-index-bar-index">
      A
    </span><span data-index="B" class="vm-index-bar-index">
      B
    </span><span data-index="C" class="vm-index-bar-index">
      C
    </span><span data-index="D" class="vm-index-bar-index">
      D
    </span><span data-index="E" class="vm-index-bar-index">
      E
    </span><span data-index="F" class="vm-index-bar-index">
      F
    </span><span data-index="G" class="vm-index-bar-index">
      G
    </span><span data-index="H" class="vm-index-bar-index">
      H
    </span><span data-index="I" class="vm-index-bar-index">
      I
    </span><span data-index="J" class="vm-index-bar-index">
      J
    </span><span data-index="K" class="vm-index-bar-index">
      K
    </span><span data-index="L" class="vm-index-bar-index">
      L
    </span><span data-index="M" class="vm-index-bar-index">
      M
    </span><span data-index="N" class="vm-index-bar-index">
      N
    </span><span data-index="O" class="vm-index-bar-index">
      O
    </span><span data-index="P" class="vm-index-bar-index">
      P
    </span><span data-index="Q" class="vm-index-bar-index">
      Q
    </span><span data-index="R" class="vm-index-bar-index">
      R
    </span><span data-index="S" class="vm-index-bar-index">
      S
    </span><span data-index="T" class="vm-index-bar-index">
      T
    </span><span data-index="U" class="vm-index-bar-index">
      U
    </span><span data-index="V" class="vm-index-bar-index">
      V
    </span><span data-index="W" class="vm-index-bar-index">
      W
    </span><span data-index="X" class="vm-index-bar-index">
      X
    </span><span data-index="Y" class="vm-index-bar-index">
      Y
    </span><span data-index="Z" class="vm-index-bar-index">
      Z
    </span></div>
  <div data-index="0">
    <div class="vm-index-anchor">
      1
    </div>
  </div>
  <div data-index="1">
    <div class="vm-index-anchor">
      2
    </div>
  </div>
  <div data-index="2">
    <div class="vm-index-anchor">
      3
    </div>
  </div>
  <div data-index="3">
    <div class="vm-index-anchor">
      4
    </div>
  </div>
</div>
`;

exports[`should update active anchor after page scroll 2`] = `
<div class="vm-index-bar">
  <div class="vm-index-bar-sidebar"><span data-index="A" class="vm-index-bar-index">
      A
    </span><span data-index="B" class="vm-index-bar-index">
      B
    </span><span data-index="C" class="vm-index-bar-index">
      C
    </span><span data-index="D" class="vm-index-bar-index">
      D
    </span><span data-index="E" class="vm-index-bar-index">
      E
    </span><span data-index="F" class="vm-index-bar-index">
      F
    </span><span data-index="G" class="vm-index-bar-index">
      G
    </span><span data-index="H" class="vm-index-bar-index">
      H
    </span><span data-index="I" class="vm-index-bar-index">
      I
    </span><span data-index="J" class="vm-index-bar-index">
      J
    </span><span data-index="K" class="vm-index-bar-index">
      K
    </span><span data-index="L" class="vm-index-bar-index">
      L
    </span><span data-index="M" class="vm-index-bar-index">
      M
    </span><span data-index="N" class="vm-index-bar-index">
      N
    </span><span data-index="O" class="vm-index-bar-index">
      O
    </span><span data-index="P" class="vm-index-bar-index">
      P
    </span><span data-index="Q" class="vm-index-bar-index">
      Q
    </span><span data-index="R" class="vm-index-bar-index">
      R
    </span><span data-index="S" class="vm-index-bar-index">
      S
    </span><span data-index="T" class="vm-index-bar-index">
      T
    </span><span data-index="U" class="vm-index-bar-index">
      U
    </span><span data-index="V" class="vm-index-bar-index">
      V
    </span><span data-index="W" class="vm-index-bar-index">
      W
    </span><span data-index="X" class="vm-index-bar-index">
      X
    </span><span data-index="Y" class="vm-index-bar-index">
      Y
    </span><span data-index="Z" class="vm-index-bar-index">
      Z
    </span></div>
  <div data-index="0">
    <div class="vm-index-anchor">
      1
    </div>
  </div>
  <div data-index="1">
    <div class="vm-index-anchor">
      2
    </div>
  </div>
  <div data-index="2">
    <div class="vm-index-anchor">
      3
    </div>
  </div>
  <div data-index="3">
    <div class="vm-index-anchor">
      4
    </div>
  </div>
</div>
`;
